ahds/grammar.py

```diff
@@ -213,6 +82,91 @@ def number_seq(self, (tag, left, right, taglist), buffer_):
 <ts>                         :=    [ \t]*
 <c>                          :=    ","
 '''
+)
+
+# dict representin structure of hypersurface file according to Amira Reference guide
+# pp 519-525 # downloaded Dezember 2018 from 
+# http://www1.udel.edu/ctcr/sites/udel.edu.ctcr/files/Amira%20Reference%20Guide.pdf
+_hyper_surface_file = {
+    'Vertices': ['Coordinates',3,'float',False],
+    'NBranchingPoints': [None,None,'int',True],
+    'NVerticesOnCurves': [None,None,'int',True],
+    'BoundaryCurves': {
+        'Vertices': [ None,1,'int',False],
+        0:True
+    },
+    'Patches': {
+        'InnerRegion':[None,None,'str',False],
+        'OuterRegion':[None,None,'str',False],
+        'Triangles':[None,3,'int',False],
+        'BranchingPoints':[None,0,'int',True],
+        'BoundaryCurves':[None,0,'int',True],
+        0:False
+    },
+    'Surfaces': {
+        'Region':[None,None,'str',False],
+        'Patches':[None,0,'int',False],
+        0:True
+    }
+}
+
+# string representing all valid keys within the above structure is inserted 
+# in the below regular expression patterns 
+_hyper_surface_entities = '|'.join([ 
+    '|'.join([_key] + ([ _vk for _vk in _dict_iter_keys(_val) if isinstance(_vk,str)] if isinstance(_val,dict) else []))
+    for _key,_val in _dict_iter_items(_hyper_surface_file)
+    if isinstance(_key,str)
+])
+
+# maximum number of bytes to be rescanned at the end of the already inspected
+# _stream_data array after new bytes have been read from the file. In case within this
+# range a data block marker (@<Num>) or any of the above HyperSurface section keys has
+# areday been successfully identified rescan starts at the byte following this match
+_rescan_overlap = max( (
+    max([len(_key)] + ([len(_vk) for _vk in _dict_iter_keys(_val) if isinstance(_vk,str)] if isinstance(_val,dict) else []))
+    for _key,_val in _dict_iter_items(_hyper_surface_file)
+    if isinstance(_key,str)
+) ) + 16
+
+if sys.version_info[0] > 2:
+    # definitions required by python3 and newer for properly parsing binary byte strings without converting
+    # them first into regular UTF-8 strings
+ 
+    _file_format_match = (
+        re.compile(b'.*AmiraMesh.*'),
+        re.compile(b'.*HyperSurface.*')
+    )
+
+    # in python3 and later open(<filename>,'rb') creates a binary file stream which has to be 
+    # explicitly decoded to unicode strings which are standard in python3 and later. Therefore
+    # any regular expression and string used to manipulate the raw stream data also has to be
+    # defined as byte string instead of regular raw pyhton string
+    _strip_lineend = b'\n'
+    _stream_delimiters = [
+        re.compile(b"(?:^|\n)@(?P<stream>\d+)\n",flags=re.S),
+        re.compile(r"(?:^|\n)\s*(?P<stream>(?:{}))(?:\s+(?:(?P<count>\d+)|(?P<name>\w+)))?(?:\s*\n|\s+)".format(_hyper_surface_entities).encode('ASCII')),
+        re.compile(b"^\s*}',re.I") # NOTE this is appried to reverese slice of stream_data therefore ^
```

Comment:
Should re.I be inside the quotes?

Reply:
No, re.I is case insensitive flag for regular expressions. I check and correct.